airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/initialsync/MySqlInitialSyncStateIterator.java

```diff
@@ -0,0 +1,98 @@
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MySqlInitialSyncStateIterator extends AbstractIterator<AirbyteMessage> implements Iterator<AirbyteMessage> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MySqlInitialSyncStateIterator.class);
+  public static final Duration SYNC_CHECKPOINT_DURATION = Duration.ofMinutes(15);
+  public static final Integer SYNC_CHECKPOINT_RECORDS = 10_000;
+
+  private final Iterator<AirbyteMessage> messageIterator;
+  private final AirbyteStreamNameNamespacePair pair;
+  private boolean hasEmittedFinalState = false;
+  private String lastPk;
+  private final JsonNode streamStateForIncrementalRun;
+  private final MySqlInitialLoadStateManager stateManager;
+  private long recordCount = 0L;
+  private Instant lastCheckpoint = Instant.now();
+  private final Duration syncCheckpointDuration;
+  private final Long syncCheckpointRecords;
+  private final String pkFieldName;
+
+  public MySqlInitialSyncStateIterator(final Iterator<AirbyteMessage> messageIterator,
+      final AirbyteStreamNameNamespacePair pair,
+      final MySqlInitialLoadStateManager stateManager,
+      final JsonNode streamStateForIncrementalRun,
+      final Duration checkpointDuration,
+      final Long checkpointRecords) {
+    this.messageIterator = messageIterator;
+    this.pair = pair;
+    this.stateManager = stateManager;
+    this.streamStateForIncrementalRun = streamStateForIncrementalRun;
+    this.syncCheckpointDuration = checkpointDuration;
+    this.syncCheckpointRecords = checkpointRecords;
+    this.pkFieldName = stateManager.getPrimaryKeyInfo(pair).pkFieldName();
+  }
+
+  @CheckForNull
+  @Override
+  protected AirbyteMessage computeNext() {
+    if (messageIterator.hasNext()) {
+      if ((recordCount >= syncCheckpointRecords || Duration.between(lastCheckpoint, OffsetDateTime.now()).compareTo(syncCheckpointDuration) > 0)
+          && Objects.nonNull(lastPk)
+          && StringUtils.isNotBlank(lastPk)) {
+        final PrimaryKeyLoadStatus pkStatus = new PrimaryKeyLoadStatus()
+            .withVersion(MYSQL_STATUS_VERSION)
+            .withStateType(StateType.PRIMARY_KEY)
+            .withPkName(pkFieldName)
+            .withPkVal(lastPk)
+            .withIncrementalState(streamStateForIncrementalRun);
+        LOGGER.info("Emitting initial sync pk state for stream {}, state is {}", pair, pkStatus);
+        recordCount = 0L;
+        lastCheckpoint = Instant.now();
+        return new AirbyteMessage()
+            .withType(Type.STATE)
+            .withState(stateManager.createIntermediateStateMessage(pair, pkStatus));
+      }
+      // Use try-catch to catch Exception that could occur when connection to the database fails
+      try {
+        final AirbyteMessage message = messageIterator.next();
+        if (Objects.nonNull(message)) {
+          lastPk = message.getRecord().getData().get(pkFieldName).asText();
```

Comment:
Cant we do the same thing that we did with Postgres and instead of trying to extract the value from the record, we just create a new class which contains the original record and then the primary key value is in separate attribute.

Reply:
IIRC the difference there was that we never wanted to emit CTID whereas here we do want to emit the PK (at least part of CDC)?

Maybe we can revisit this when we build this out for the cursor based sync and the PK might not always be emitted - though in general I do think it would be a weird use case to omit the PK all together